trinity/protocol/common/managers.py

```diff
@@ -214,3 +237,129 @@ def _send_sub_proto_request(self, request: TRequest) -> None:
+        validate the received message. If no ValidationError is raised, then the
+        message will be considered delivered, and the response closed.
+        """
+        self._request(make_request)
+        while self._is_pending():
+            yield await self._get_message(timeout)
+
+    @contextmanager
+    def complete_if_valid(self, item_count: int) -> Generator[None, None, None]:
+        try:
+            yield
+        except ValidationError as err:
+            self.logger.debug(
+                "Response validation failed for pending %s request from peer %s: %s",
+                self.response_msg_name,
+                self._peer,
+                err,
+            )
+        else:
+            self.pending_request = None
+            self.response_times.add(self.last_response_time, item_count)
+
+    @property
+    def response_msg_name(self) -> str:
+        return self._response_msg_type.__name__
+
+    #
+    # Service API
+    #
+    async def _run(self) -> None:
+        self.logger.debug("Launching %s for peer %s", self.__class__.__name__, self._peer)
+
+        with self.subscribe_peer(self._peer):
+            while self.is_running:
+                peer, cmd, msg = await self.wait(self.msg_queue.get())
+                if peer != self._peer:
+                    self.logger.error("Unexpected peer: %s  expected: %s", peer, self._peer)
+                    continue
+                elif isinstance(cmd, self._response_msg_type):
+                    await self._handle_msg(cast(TMsg, msg))
+                else:
+                    self.logger.warning("Unexpected message type: %s", cmd.__class__.__name__)
+
+    async def _handle_msg(self, msg: TMsg) -> None:
+        if self.pending_request is None:
+            self.logger.debug(
+                "Got unexpected %s message from %", self.response_msg_name, self._peer
+            )
+            return
+
+        send_time, future = self.pending_request
+        self.last_response_time = time.perf_counter() - send_time
+        future.set_result(msg)
+
+    async def _get_message(self, timeout: int) -> TMsg:
+        send_time, future = self.pending_request
+        try:
+            message = await self.wait(future, timeout=timeout)
+        except TimeoutError:
+            self.response_times.total_timeouts += 1
+            self.pending_request = None
+            raise
+        else:
+            # message might be invalid, so allow another attempt to wait for a valid response
+            self.pending_request = (send_time, asyncio.Future())
+
+        return message
+
+    def _request(self, make_request: Callable[[], None]) -> None:
+        if self.pending_request is not None:
+            self.logger.error(
+                "Already waiting for response to %s for peer: %s",
+                self.response_msg_name,
+                self._peer,
+            )
+            raise AlreadyWaiting(
```

Comment:
orthogonal to this pull request, but I've been puzzling over how we can eliminate this failure condition with either a `asyncio.Lock` or `asyncio.Condition` to allow call sites to *aquire* the right to issue a request, and *then* issue it with a strong guarantee that the only errors they are likely to encounter are a `TimeoutError`.  I haven't actually tried anything but I wanted to put that out there in case anyone has any bright ideas.